### PR TITLE
PHOENIX-5734 - IndexScrutinyTool should not report rows beyond maxLoo…

### PR DESCRIPTION
**Pull request by gjacoby126**
…kBack age

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolBaseIT.java [hunk 1]
```diff
@@ -54,8 +54,9 @@
                 new ReadOnlyProps(clientProps.entrySet().iterator()));
     }
 
-    protected List<Job> runScrutiny(String[] cmdArgs) throws Exception {
-        IndexScrutinyTool scrutiny = new IndexScrutinyTool(IndexScrutinyMapperForTest.class);
+    protected List<Job> runScrutiny(Class<? extends IndexScrutinyMapper> mapperClass,
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyToolBaseIT.java]: aah! nice :)

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    private static String dataTableName;
+    private static String indexTableName;
+    private static String viewName;
+    private static boolean isViewIndex;
+    private static ManualEnvironmentEdge testClock;
+    public static final String UPSERT_DATA = "UPSERT INTO %s VALUES (?, ?, ?)";
+
+    @BeforeClass
+    public static synchronized void doSetup() throws Exception {
+        Map<String, String> props = Maps.newHashMapWithExpectedSize(1);
```
Comment (gokceni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: nit : consider expected size = 2

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        props.put(ScanInfoUtil.PHOENIX_MAX_LOOKBACK_AGE_CONF_KEY,
+            Integer.toString(MAX_LOOKBACK));
+        setUpTestDriver(new ReadOnlyProps(props.entrySet().iterator()));
+    }
+
+    @Test
+    public void testScrutinyOnRowsBeyondMaxLookBack() throws Exception {
+        schema = generateUniqueName();
+        dataTableName = generateUniqueName();
+        indexTableName = generateUniqueName();
```
Comment (gokceni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: nit: consider prepending "I_" and "D_" to names

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            // insert two rows
+            upsertDataStmt = getUpsertDataStmt(tableFullName, conn);
+
+            NonParameterizedIndexScrutinyToolIT.upsertRow(upsertDataStmt, 1, "name-1", 98051);
+            NonParameterizedIndexScrutinyToolIT.upsertRow(upsertDataStmt, 2, "name-2", 98052);
+            conn.commit();
+            long afterInsertSCN = EnvironmentEdgeManager.currentTimeMillis() + 1;
+            testClock.setValue(afterInsertSCN);
+            EnvironmentEdgeManager.injectEdge(testClock);
+            testClock.incrementValue(1);
```
Comment (gokceni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: nit: Would be good to have a comment on what the clock increments are doing.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            long scrutinyTs = EnvironmentEdgeManager.currentTimeMillis();
+            changeIndexValue();
+            testClock.incrementValue(MAX_LOOKBACK /2  * 1000);
+            List<Job> completedJobs = runScrutiny(schema, tableName, indexTableName, scrutinyTs);
+            Job job = completedJobs.get(0);
+            assertTrue(job.isSuccessful());
+            assertCounters(job.getCounters());
+        }
+    }
+    private void changeIndexValue() throws SQLException {
```
Comment (gokceni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: Suggest to rename this function to be add index row

File: phoenix-core/src/test/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapperForTest.java [hunk 6]
```diff
@@ -23,6 +23,8 @@
 public class IndexScrutinyMapperForTest extends IndexScrutinyMapper {
 
     public static final int TEST_TABLE_TTL = 3600;
+    public static final int MAX_LOOKBACK = 6;
```
Comment (swaroopak) [inline phoenix-core/src/test/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapperForTest.java]: nit: unnecessary change

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java [hunk 7]
```diff
@@ -288,8 +301,15 @@ protected void checkIfInvalidRowsExpired(Context context,
             Pair<Long, List<Object>> sourceValues = entry.getValue();
             Long sourceTS = sourceValues.getFirst();
             if (hasRowExpiredOnSource(sourceTS, ttl)) {
-                context.getCounter(PhoenixScrutinyJobCounters.EXPIRED_ROW_COUNT).increment(1);
-                itr.remove();
+                context.getCounter(PhoenixScrutinyJobCounters.EXPIRED_ROW_COUNT).increment(1L);
+                itr.remove(); //don't output to the scrutiny table
+            } else if (isRowOlderThanMaxLookback(sourceTS)){
+                context.getCounter(PhoenixScrutinyJobCounters.BEYOND_MAX_LOOKBACK).increment(1L);
+                //still output to the scrutiny table just in case it's useful
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java]: why do we need them around?

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java [hunk 8]
```diff
@@ -299,7 +319,16 @@ protected boolean hasRowExpiredOnSource(Long sourceTS, Long ttl) {
         return ttl != Integer.MAX_VALUE && sourceTS + ttl*1000 < currentTS;
     }
 
-    private long getTableTtl() throws SQLException, IOException {
+    protected boolean isRowOlderThanMaxLookback(Long sourceTS){
```
Comment (swaroopak) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java]: nit: space before '{' this and some other places.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    private static class MaxLookbackIndexScrutinyMapper extends IndexScrutinyMapper {
+        @Override
+        public void postSetup(){
+            try {
+                String tableToCompact;
+                if (isViewIndex){
+                    String physicalDataTableName =
+                        SchemaUtil.getPhysicalHBaseTableName(schema, dataTableName, false).getString();
```
Comment (swaroopak) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: why not take isNamespaceEnabled property from the config?

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java [hunk 10]
```diff
@@ -161,12 +168,22 @@ protected void setup(final Context context) throws IOException, InterruptedExcep
...
         } catch (SQLException | NoSuchAlgorithmException e) {
             tryClosingResourceSilently(this.outputUpsertStmt);
             tryClosingResourceSilently(this.connection);
             tryClosingResourceSilently(this.outputConn);
             throw new RuntimeException(e);
         }
+        postSetup();
+    }
+
+    protected void preSetup() {
```
Comment (gokceni) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java]: Any reasons to have these empty methods?

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTool.java [hunk 11]
```diff
@@ -505,6 +509,18 @@ public int run(String[] args) throws Exception {
         }
     }
 
+    private void validateTimestamp(Configuration configuration, long ts) {
+        long maxLookBackAge = ScanInfoUtil.getMaxLookbackInMillis(configuration);
+        if (maxLookBackAge != ScanInfoUtil.DEFAULT_PHOENIX_MAX_LOOKBACK_AGE * 1000L) {
```
Comment (gokceni) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTool.java]: Is there a chance that people coincidentally pick the same Max lookback age as the default one?

Comment (gjacoby126) [reply phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTool.java]: The default age is the same as "turn off max lookback". It happens to be 0, which makes the 1000L not necessary, but I wanted to keep the discipline of _always_ converting to millis from seconds to avoid accidentally missing it elsewhere.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java (hunk 7 above)
Comment (gjacoby126) [reply phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java]: Humans need to evaluate beyond max lookback counters to see if they're valid issues or not.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+            conn.createStatement().execute(String.format(dataTableDDL, dataTableFullName));
+            conn.createStatement().execute(String.format(viewDDL, viewFullName, dataTableFullName));
+            conn.createStatement().execute(String.format(indexTableDDL, indexTableName,
+                viewFullName));
+            conn.commit();
+        }
+        upsertDataAndScrutinize(viewName, viewFullName, testClock);
+    }
+
+    private void upsertDataAndScrutinize(String tableName, String tableFullName,
```
Comment (kadirozde) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: Can we have tests for deletes in addition to upserts? I am also interested in the cases where rows are deleted within the max loopback window and before and after the scn.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/MaxLookbackIT.java [hunk 13]
```diff
@@ -317,16 +300,16 @@ public void testRecentMaxVersionsNotCompactedAway() throws Exception {
...
             //after major compaction, check to make sure we can see all three versions
             // at the appropriate times
             assertMultiVersionLookbacks(dataTableSelectSql, allValues, allSCNs);
             assertMultiVersionLookbacks(indexTableSelectSql, allValues, allSCNs);
             injectEdge.incrementValue(MAX_LOOKBACK_AGE * 1000);
             long afterLookbackAgeSCN = EnvironmentEdgeManager.currentTimeMillis();
-            majorCompact(dataTable, afterLookbackAgeSCN);
-            majorCompact(indexTable, afterLookbackAgeSCN);
+            majorCompact(dataTable);
+            majorCompact(indexTable);
```
Comment (kadirozde) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/MaxLookbackIT.java]: In the tests above, major compactions run on both the data and index table. Can we pick only one table for one test, the other for another test, and so on? The interesting problems arise when compaction runs on one table but not on both

Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/MaxLookbackIT.java]: First, this test suite only was modified because I needed to steal its major compaction code and put it into TestUtil. It's otherwise unrelated to this JIRA, and if it needs changes, that should be a separate issue. 

Second, this isn't really a test of indexing per se, but of whether the max lookback coproc hooks correctly prevent data from being purged during the lookback window.

Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/MaxLookbackIT.java]: But if there are additional indexing tests we need, definitely please suggest them in a JIRA!

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java (hunk 12 above)
Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: Good idea, I will add a test for deletes.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/MaxLookbackIT.java (hunk 13 above)
Comment (kadirozde) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/MaxLookbackIT.java]: That is fair. Thanks.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixScrutinyJobCounters.java [hunk 14]
```diff
@@ -41,5 +41,11 @@
...
      * Number of batches processed
      */
-    BATCHES_PROCESSED_COUNT;
+    BATCHES_PROCESSED_COUNT,
+    /**
+     * Number of rows in source that became older than the max lookback age while scrutiny
+     * was comparing them with the target, and didn't match. We break these out separately because
+     * they could be due to extra versions being compacted, and are harmless.
+     */
+    BEYOND_MAX_LOOKBACK;
```
Comment (priyankporwal) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixScrutinyJobCounters.java]: All counters end with '_COUNT' here ... Perhaps we should remain consistent. That said, 'BEYOND_MAX_LOOKBACK_ROW_COUNT' looks too long. How about 'NO_LOOKBACK_ROW_COUNT'?

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java [hunk 15]
```diff
@@ -1183,6 +1186,22 @@ public static int getIsNullableInt(boolean isNullable) {
...
+    }
+
+    public static int getTimeToLive(PhoenixConnection conn, byte[] tableQualifier)
+     throws SQLException {
+        HTableDescriptor td = conn.getQueryServices().getTableDescriptor(tableQualifier);
+        HColumnDescriptor[] cds = td.getColumnFamilies();
+        if (cds.length > 0){
+            return cds[0].getTimeToLive();
+        } else {
+            return HConstants.FOREVER;
```
Comment (priyankporwal) [inline phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java]: when would this be the case? i.e. #CF==0

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixScrutinyJobCounters.java (hunk 14 above)
Comment (gjacoby126) [reply phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixScrutinyJobCounters.java]: No objection to appending _COUNT to the existing enum value, but NO_LOOKBACK_ROW_COUNT seems more confusing to me.

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java (hunk 15 above)
Comment (gjacoby126) [reply phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java]: I couldn't think of one, but decided to code the function defensively.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixScrutinyJobCounters.java (hunk 14 above)
Comment (priyankporwal) [reply phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/PhoenixScrutinyJobCounters.java]: True. They are similar to 'EXPIRED' but can't use the same name... 'UNREADABLE' can also be confusing. Anyways, pick what sounds best to you.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java (hunk 7 above)
Comment (priyankporwal) [reply phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java]: @gjacoby126 Any reason to be concerned about bloating the scrutiny table? This is an expected race condition that can happen pretty regularly if rows are updated at the cadence of around MaxLookbackAge. Humans may not necessarily have bandwidth to look at potentially large number of rows. If we are going to emit to scrutiny table, perhaps we should add a way to filter these out as well to make them manageable.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java (hunk 9 above)
Comment (swaroopak) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: @gjacoby126

Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexScrutinyWithMaxLookbackIT.java]: I know it's going to be false.

File: phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTableOutput.java [hunk 16]
```diff
@@ -68,14 +68,18 @@
...
+            "    BEYOND_MAX_LOOKBACK BOOLEAN,\n" +
             "    CONSTRAINT PK PRIMARY KEY\n" +
             "    (\n" +
             "        " + SOURCE_TABLE_COL_NAME + ",\n" +
             "        " + TARGET_TABLE_COL_NAME + ",\n" +
             "        " + SCRUTINY_EXECUTE_TIME_COL_NAME + ",\n" + // time at which the scrutiny ran
             "        SOURCE_ROW_PK_HASH\n" + //  this hash makes the PK unique
             "    )\n" + // dynamic columns consisting of the source and target columns will follow
-            ")";
+            ")  COLUMN_ENCODED_BYTES = 0 "; //column encoding not supported with dyn columns (PHOENIX-5107)
```
Comment (priyankporwal) [inline phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyTableOutput.java]: Do we need to do anything to disable COLUMN_ENCODIG for clusters with this table already existing?

**Comment by gjacoby126** [comment]
Thanks for the reviews @swaroopak @priyankporwal @kadirozde
